File: success/46.py
class Solution:
    def __init__(self):
        self.permutations = []
        self.permutation = []
        
    def permute(self, nums: List[int]) -> List[List[int]]:
        visited = [False for _ in range(len(nums))]
        self.recursive_approach(nums, visited)
        return self.permutations
        
    def recursive_approach(self, nums, visited):

        if len(self.permutation) == len(nums):
            self.permutations.append(self.permutation[:])
            return
        
        for i in range(len(nums)):
            if not visited[i]:
                
                self.permutation.append(nums[i])
                visited[i] = True
                
                self.recursive_approach(nums, visited)
                
                self.permutation.pop()
                visited[i] = False

File: success/46.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<bool> visited;
        visited.assign(nums.size(), false);
        RecursiveApproach(nums, visited);
        return permutations;
    }
    
private:
    vector<vector<int>> permutations;
    vector<int> permutation;
    
    void RecursiveApproach(vector<int>& nums, vector<bool>& visited) {
        
        if (permutation.size() == nums.size()) {
            permutations.push_back(permutation);
            return;
        }
        
        for (auto i = 0u; i < nums.size(); i++) {
            if (!visited[i]) {
                
                permutation.push_back(nums[i]);
                visited[i] = true;
                
                RecursiveApproach(nums, visited);
                
                permutation.pop_back();
                visited[i] = false;
            }
        }
    }
};